Introduce ComputedUrl

fnune opened the pull request:
<!-- Please describe your changes on the following line: -->

Use the new `ComputedUrl` type for computed types and `SpecifiedUrl` for specified types instead of using the `SpecifiedUrl` implementation for both.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #17625 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because this is an implementation change and tests already exist.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17812)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/servo/url.rs
 * @canaltinova: components/style/servo/url.rs
 * @emilio: components/style/servo/url.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

jdm commented:
r? @nox

emilio commented:
Yeah, this is not the right patch and I bet it fails tests.

What needs to happen is that this piece of code needs to go away, at least for Servo:

```rust
// TODO(emilio): Maybe consider ComputedUrl to save a word in style structs?
impl ComputedValueAsSpecified for SpecifiedUrl {}
```

And instead we need to make something like:

```rust
enum ComputedUrl {
    Invalid(Arc<String>),
    Valid(ServoUrl),
}
```

And `impl ToComputedValue for SpecifiedUrl` properly, and also `ToCss`.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 62bccc23a2af941e909f968c8bda12f9742c38bd with merge afdc1f3220380d402cf11700353597b32ad0dfff...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5146)

jdm commented:
```
  ▶ Unexpected subtest result in /_mozilla/css/import_serialization.html:
  │ FAIL [expected PASS] CSS Test: @import rules are correctly serialized
  │   → assert_equals: expected "@import url(\"foo.css\");" but got "@import url(\"http://web-platform.test:8000/_mozilla/css/foo.css\");"
  │ 
  │ @http://web-platform.test:8000/_mozilla/css/import_serialization.html:13:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/_mozilla/css/import_serialization.html:11:1
```

brainlessdeveloper commented:
Yep seems pretty clear. Would the new struct need to be used anywhere else?

emilio commented:
I don't think so, for now at least.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17822) made this pull request unmergeable. Please resolve the merge conflicts.

brainlessdeveloper commented:
Working on this now!

brainlessdeveloper commented:
What should `SpecifiedUrl::to_css` return if there is no `original`? Because after these changes, `SpecifiedUrl` should not try to access `resolved`, right?

fnune committed:
Introduce ComputedUrl

Add web platform tests for computed URL styles

Mark url with no original or resolved unreachable

Update the WPT manifest for new url tests

brainlessdeveloper commented:
I'd like some input in the current progress. Also, currently I'm trying to understand some (to me) quite cryptic error messages during compilation, which I think are related to the associated type `ComputedValue` in the implementation of `ToComputedValue for SpecifiedUrl`:

```
error[E0308]: mismatched types                 
   --> /home/fausto/Dev/rust/servo/components/style/values/generics/image.rs:160:35
    |                                          
160 | #[derive(Clone, Debug, PartialEq, ToComputedValue, ToCss)]
    |                                   ^^^^^^^^^^^^^^^ expected struct `servo::url::SpecifiedUrl`, found enum `values::computed::ComputedUrl`
    |                                          
    = note: expected type `servo::url::SpecifiedUrl`                                          
               found type `values::computed::ComputedUrl`                                     

error[E0271]: type mismatch resolving `<servo::url::SpecifiedUrl as values::computed::ToComputedValue>::ComputedValue == servo::url::SpecifiedUrl`
   --> /home/fausto/Dev/rust/servo/components/style/values/generics/image.rs:160:35
    |                                          
160 | #[derive(Clone, Debug, PartialEq, ToComputedValue, ToCss)]
    |                                   ^^^^^^^^^^^^^^^ expected enum `values::computed::ComputedUrl`, found struct `servo::url::SpecifiedUrl`
    |                                          
    = note: expected type `values::computed::ComputedUrl`                                     
               found type `servo::url::SpecifiedUrl` 
```

Unfortunately I don't know how to go on about debugging those :c

emilio commented:
> I'd like some input on the current progress. Also, currently I'm trying to understand some (to me) quite cryptic error messages during compilation, which I think are related to the associated type ComputedValue in the implementation of ToComputedValue for SpecifiedUrl:

Hi! The work so far looks quite reasonable :)

The errors you're seeing is because `values::generic::Image` is supposed to serve both as a computed image and a specified image, depending on the template parameters. That is, the `ToComputedValue` implementation does something like:

```rust
impl<Gradient, MozImageRect> ToComputedValue for Image<Gradient, MozImageRect>
where
    Gradient: ToComputedValue,
    MozImageRect: ToComputedValue,
{
    type ComputedValue = Image<Gradient::ComputedValue, MozImageRect::ComputedValue>;

    fn to_computed_value(&self, context: &Context) -> Self::ComputedValue {
        match *self {
            Image::Url(ref url) => Image::Url(url.to_computed_value(context)),
            Image::Gradient(ref gradient) => Image::Gradient(gradient.to_computed_value(context)),
            // ...
        }            
    }
}
```

As you may see, now that `url.to_computed_value(context)` returns something that isn't a `SpecifiedUrl`, the `Image::Url` value would be invalid.

Basically, you need to make `Image` generic also on the `Url` type, so it can be `SpecifiedUrl` when used as a specified value, and `ComputedUrl` when used as a computed value.

Hope it helps!

emilio commented:
> What should SpecifiedUrl::to_css return if there is no original? Because after these changes, SpecifiedUrl should not try to access resolved, right?

Why not? We only won't have `specified` using the uncompute functionality, and in that case it's fine to just use the resolved

brainlessdeveloper commented:
That does help a lot! Thanks. I'm making progress but there's a problem after some changes I've made to a file that uses templating (`.mako.rs`). After a failed build, a consecutive attempt will fail earlier and report some error in an `out` directory, and I have to remove the build folder and start again:
```
 servo git:(use-resolved-url-instead-of-original) ✗ ./mach build -d
error: failed to read `/home/fausto/Dev/rust/servo/target/debug/build/osmesa-src-61979e1c0efa04c9/output`

Caused by:
  No such file or directory (os error 2)
```
I can continue making changes like this but it's pretty annoying, am I missing anything?

emilio commented:
So... That's pretty weird, seems like the osmesa build got messed up somehow...

You can probably delete the `osmesa-src-<hash>` directory, I'd expect that to restart the build from scratch, so it should work.

If that doesn't work, you can probably delete https://github.com/servo/servo/blob/master/ports/glutin/Cargo.toml#L42-L56, since osmesa is only there to provide the mesa runtime when running tests.

brainlessdeveloper commented:
I see, I've had errors like that come up in different builds, for example right now in `target/debug/build/style-c1fe248c3e80ee66/out/properties.rs:51081:33`. I'll see if I can make progress faster with your recommendation, thanks.

emilio commented:
Oh, yeah, the errors in `properties.rs` are annoying, but they come from the multiple `.mako.rs` files. Usually it's not hard to open the autogenerated file and track down which of the `longhands/foo.mako.rs` or `shorthands/foo.mako.rs` files it comes from.

brainlessdeveloper commented:
So, I got it to compile but the tests aren't passing yet. Current status: I'm getting errors in files generated by `mako` templates. Example:
```
error[E0308]: mismatched types                                                
     --> /home/fausto/Dev/rust/servo/target/debug/build/style-9e8aa6a6aa746743/out/properties.rs:45265:66
      |                                                                               
45265 |                                 context.builder.set_marker_start(computed)                       
      |                                                                  ^^^^^^^^ expected struct `servo::url::SpecifiedUrl`, found enum `values::computed::ComputedUrl`
      |                                                                                                  
      = note: expected type `values::Either<servo::url::SpecifiedUrl, _>`                                                              
                 found type `values::Either<values::computed::ComputedUrl, _>`
```
I guess I should make structs like this:
```
<%helpers:shorthand name="marker" products="gecko"
    sub_properties="marker-start marker-end marker-mid"
    spec="https://www.w3.org/TR/SVG2/painting.html#MarkerShorthand">
    use values::specified::UrlOrNone;
```
...generic over the `Url` type. But I'm not too sure how bcause `UrlOrNone` is defined in the `specified` directory, implying it shouldn't be "specified/computed url or none".

Anyway, I'll be on this again tomorrow evening if I get some free time :)

emilio commented:
That's because in `values/computed/mod.rs` there's a line that reads:

```rust
pub use super::specified::{BorderStyle, UrlOrNone};
```

But that's not true anymore, since `computed::UrlOrNone` shouldn't be just `specified::UrlOrNone`. So you'll have to remove `UrlOrNone` from there, and add something like:

```rust
pub type UrlOrNone = Either<ComputedUrl, None_>
```

I believe that should be enough to fix that compiler error.

Thanks again for working on this, it's great!

brainlessdeveloper commented:
Previously failing `"CSS Test: @import rules are correctly serialized"` from WPT now passes but I haven't tested the entire suite. r? @emilio

brainlessdeveloper commented:
This fails on `./mach build-geckolib`:

```
15 | use servo_url::ServoUrl;                              
   |     ^^^^^^^^^^^^^^^^^^^ Maybe a missing `extern crate servo_url;`? 
```

And exactly the same on `./mach test-stylo`. I suppose it's because of these `cfg`s:
```
#[cfg(feature = "servo")] #[macro_use] extern crate serde; 
#[cfg(feature = "servo")] extern crate servo_url;
```

Should I just remove the `feature` so it compiles for gecko too or should I do it some other way?

emilio commented:
Hmm... I think we want ComputedValueAsSpecified for Gecko's URLs, for now, and ComputedUrl being a servo-only type. It's somewhat annoying though.

brainlessdeveloper commented:
Alright I'll try to fix it later, thanks!

brainlessdeveloper commented:
Alright, now everything passed on my machine, except for `wpt`, which I haven't run yet. @emilio r?

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 8dc9e338a83ec9aa8acf4ad37f4923a3033d1b7c with merge 9d7a3c6fad485ddee65ec998cae4df54311b05d3...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/873)

brainlessdeveloper commented:
I don't know what to do about a failure on the Mac suite.

KiChjang commented:
@bors-servo retry
* infra phail

bors-servo commented:
:hourglass: Trying commit 8dc9e338a83ec9aa8acf4ad37f4923a3033d1b7c with merge ab0582ed7fc66c893be3a9d0b1c966d3822f92bd...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8054), [arm32](http://build.servo.org/builders/arm32/builds/8277), [arm64](http://build.servo.org/builders/arm64/builds/8310), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8396), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5157), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5190), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8391), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/868), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/867), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5093), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5151), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/869), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/874), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3635)
State: approved= try=True

brainlessdeveloper commented:
r?

emilio commented on components/style/values/computed/mod.rs:
```diff
@@ -586,3 +589,45 @@ impl ClipRectOrAuto {
+/// The computed value of a CSS `url()`, resolved relative to the stylesheet URL.
+#[cfg(feature = "servo")]
+#[derive(Clone, Debug, HeapSizeOf, Serialize, Deserialize, PartialEq)]
+pub enum ComputedUrl {
```
We could move this to another module to avoid the `cfg`'s, but that's probably doable as a followup.

brainlessdeveloper replied:
You mean I should create a new files like this:
```
computed/ur/servo.rs
computed/url/gecko.rs
computed/url/mod.rs // Export here? Would still need the cfgs am I right?
```

emilio replied:
Yeah, I meant this, but I wouldn't worry about that for this PR, it's fine to do it in followup work.

emilio commented on components/style/servo/url.rs:
```diff
@@ -126,3 +127,25 @@ impl ToCss for SpecifiedUrl {
+    fn to_computed_value(&self, _: &Context) -> Self::ComputedValue {
+        match self.resolved {
+            Some(ref url) => ComputedUrl::Valid(url.clone()),
+            None => ComputedUrl::Invalid(Arc::new("about:invalid".to_string())),
```
Why `about:invalid`, and not the specified one? Off-hand, that should be both cheaper and more correct.

What do other browsers do here?

brainlessdeveloper replied:
Sounds like a better idea.
I'll research other browsers this afternoon, I'll be in the Mozilla offices here in Berlin and I can probably ask directly :D

emilio commented on components/style/servo/url.rs:
```diff
@@ -126,3 +127,25 @@ impl ToCss for SpecifiedUrl {
+        }
+    }
+
+    fn from_computed_value(computed: &ComputedUrl) -> Self {
```
Also, if you did what I suggest above, you could use `original: None, resolved: Some()` for valid urls, and `original: Some(), resolved: None` for invalid ones, which would match what we do when computing them.

brainlessdeveloper replied:
Agree here too.

emilio reviewed (commented):
Looks pretty good! Just some questions.

brainlessdeveloper commented on components/style/values/computed/mod.rs:
```diff
@@ -586,3 +589,45 @@ impl ClipRectOrAuto {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        let string = match *self {
+            ComputedUrl::Valid(ref url) => url.as_str(),
+            ComputedUrl::Invalid(ref invalid_string) => invalid_string,
```
With the new changes this `invalid_string` should either be the original URL if it was specified, or `"about:invalid"` if not specified.

brainlessdeveloper commented:
r?

emilio commented on components/style/servo/url.rs:
```diff
@@ -126,3 +127,33 @@ impl ToCss for SpecifiedUrl {
+            Some(ref url) => ComputedUrl::Valid(url.clone()),
+            None => match self.original {
+                Some(ref url) => ComputedUrl::Invalid(url.clone()),
+                None => ComputedUrl::Invalid(Arc::new("about:invalid".to_string())),
```
I don't think this can be reached (and it'd be tricky because `about:invalid` is a valid URL. I think we should be able to mark this `unreachable!("Found specified url with neither resolved or original  URI!")`, or something like that.

brainlessdeveloper replied:
I think this is the same case as the `ToCss` implementation for `SpecifiedUrl`, when there's no original and no resolved:
```
// This can only happen if the url wasn't specified by the
// user *and* it's an invalid url that has been
// back to specified value via the "uncompute" functionality.
```
So I think the case is reachable, but I'm not sure the behavior is appropriate.

emilio replied:
Right, but with your patch we're changing that, right? And the uncompute stuff (which is the `from_computed_value` implementation) now will always have one or the other.

brainlessdeveloper replied:
Yeap you're right. That was obvious... in hindsight :) I'll take care of it.

emilio reviewed (approved):
Looks great, thanks!

Please double-check me in the "can't be reached" path, and let's get this merged! :)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit d268cdf246d8ee4907c026f82028bd69847ca094 with merge 093d18f81910713ad39cd62d38893d8918237789...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8131), [arm32](http://build.servo.org/builders/arm32/builds/8354), [arm64](http://build.servo.org/builders/arm64/builds/8387), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8473), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5235), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5269), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8468), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/945), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/944), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5170), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5228), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/947), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/951), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3712)
State: approved= try=True

emilio commented:
Also, can we add a test for this, given none are written?

It'd be something like:

```
<div style="background-image: url(relative.png);"></div>
```

And assert that getComputedStyle(div).backgroundImage returns the resolved value, and div.style.backgroundImage the unresolved one.

brainlessdeveloper commented:
I wanted to write one like that but couldn't figure out where/how exactly. I've never written tests for servo. Where should I place it, and can I find any existing examples which are similar?

emilio commented:
So you can write a new test as described in: https://github.com/servo/servo/blob/master/tests/wpt/README.md.

What you want is probably a very simple test setting the background-image of an element and checking `element.style.backgroundImage` and `getComputedStyle(element).backgroundImage` for differences.

Since it's hard to find a spec that backs it out, even though other browsers seem to do the same, we can put it in `tests/wpt/mozilla/tests/css`.

brainlessdeveloper commented:
So, writing the tests was a good idea (duh). Apparently the `list-style-image` property and the shorthand `list-style` still use the specified URL. I'm onto fixing that.

Also, I can't manage to write tests for `content: url(test.jpg);` because I can't get `getComputedStyle(elem, "::after");` to return my styles. If I create a `div` with an `::after` and set `content: url(test.jpg);` on it, `getComputedStyle(elem, "::after");` will return `normal` which is not what I set. I haven't managed to work around this.

brainlessdeveloper commented:
So, the trouble with `list-style-image` is [here](https://github.com/servo/servo/blob/4f0821192c112943bb53b4fb04303c1afdde06e6/components/style/properties/longhand/list.mako.rs#L103) but I can't figure out how to make the type generic for the URL... I'm getting really confused with the templating.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17924) made this pull request unmergeable. Please resolve the merge conflicts.

brainlessdeveloper commented:
Sorry about the terrible force pushes, I messed up the rebase because it introduced compile errors.

brainlessdeveloper commented:
So to fix the new WPT case I need to make the implementations [here](https://github.com/servo/servo/blob/4f0821192c112943bb53b4fb04303c1afdde06e6/components/style/properties/longhand/list.mako.rs#L103) generic on the URL type. Right now there's some code reuse going on with `ComputedValueAsSpecified`. The problem is I can't figure out how to make my changes in the template, I don't understand where I can introduce type parameters or whether I should rebuild the property somewhere else.

Any ideas?

jdm commented:
You may need to introduce a [UrlOrNone](https://doc.servo.org/style/values/specified/type.UrlOrNone.html) type that uses computed URLs; you could then use that for the computed value, and use the original UrlOrNone type forthe specified value and remove the ComputedValueAsSpecified logic.

Note; these are guesses. If they sound sensible, give it a try while you're waiting for an answer from someone more knowledgeable ;)

brainlessdeveloper commented:
Yep, I've already implemented a `UrlOrNone` type for computed URLs, but I think the problem is `parse` should be generic over the URL type and I can't add a type parameter to it because the `impl` block is implicit because of the template stuff (I think?).

jdm commented:
When does `parse` need to return a computed URL value?

jdm commented:
Keep in mind that the `pub fn parse` is just a function inside of a `list_style_image` module; it's not a method that's part of an impl.

brainlessdeveloper commented:
I can't figure out how to make the tests pass. I don't understand how the template translates to the actual implementation.

On a side note, I just moved and I won't have an internet connection at home until next week, so I won't be able to work on this until then.

jdm commented:
To see the actual implementation, open up target/debug/build/style-[hash]/out/properties.rs and search for `mod list_style_image`.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17783) made this pull request unmergeable. Please resolve the merge conflicts.

brainlessdeveloper commented:
I won't be able to work during this week so I've [updated the test expectations for `listImage` properties](https://github.com/servo/servo/pull/17812/files#diff-b4e4af6248b356d099b97b18764bd91aR1). Is it fine if we try to merge this and I create an issue for the missing functionality?

The branch should pass the WPT as it is.

emilio commented on components/style/values/animated/mod.rs:
```diff
@@ -13,13 +13,15 @@ use smallvec::SmallVec;
 use values::computed::NonNegativeLengthOrPercentage as ComputedNonNegativeLengthOrPercentage;
 use values::computed::NonNegativeNumber as ComputedNonNegativeNumber;
 use values::computed::PositiveInteger as ComputedPositiveInteger;
+#[cfg(feature = "servo")]
```
You should remove this `cfg()`, afaict, for `./mach build-geckolib` to pass... or moving it to the `ComputedUrl`, probably.

emilio reviewed (commented):
It looks reasonable to me, this is pretty neat work!

Let's:

 * File a followup issue (happy to do so, though you may have also a bit more context :)). Also, let's reference it from a comment near the relevant place.
 * Fix the build (which is failing in Travis, and which fix I think is a one-liner I suggest below)

And then  and get this landed, so it doesn't bitrot.

Thanks a lot again!

brainlessdeveloper commented:
#18015

emilio commented on components/style/values/computed/mod.rs:
```diff
@@ -39,9 +43,8 @@ pub use self::image::{Gradient, GradientItem, Image, ImageLayer, LineDirection,
 pub use self::rect::LengthOrNumberRect;
 pub use super::{Auto, Either, None_};
-pub use super::specified::{BorderStyle, UrlOrNone};
+pub use super::specified::{BorderStyle};
```
nit: I'm not sure if the lint catches this, but we usually don't wrap in braces single imports.

emilio commented on components/style/values/computed/mod.rs:
```diff
@@ -645,3 +648,45 @@ impl From<Au> for NonNegativeAu {
+    }
+}
+
+#[allow(missing_docs)]
```
nit: could you use `<url> | <none>` as the doc comment?

emilio commented on tests/wpt/mozilla/meta/css/get-computed-style-for-url.html.ini:
```diff
@@ -0,0 +1,9 @@
+  type: testharness
+
+  [getComputedStyle(elem) for url() listStyle uses the resolved URL and elem.style uses the original URL]
+    expected: FAIL
```
Let's add a `bug: https://github.com/servo/servo/issues/18015` line here and below.

emilio commented on tests/wpt/mozilla/tests/css/get-computed-style-for-url.html:
```diff
@@ -0,0 +1,49 @@
+
+  // Some shorthands like background and border-image compute to other properties
+  var computedPropertyName = {
+    background: "backgroundImage",
```
You could just set `backgroundImage`, etc instead... though not a big deal.

brainlessdeveloper replied:
Yeap, I'm doing that as well, but I wanted to test the shorthand and the longhand too.

emilio commented on tests/wpt/mozilla/tests/css/get-computed-style-for-url.html:
```diff
@@ -0,0 +1,49 @@
+  }
+  function testUrlsForProperty(property, urlValue, extraShorthand) {
+    test(function() {
+      var extra = extraShorthand ? ` ${extraShorthand}` : "";
```
nit: `extraShorthand ? extraShorthand.toString() : ""`?

emilio replied:
(But setting the longhand property directly would allow you to avoid this complexity).

brainlessdeveloper replied:
The problem is the space before the `extraShorthand`, which I add using the template string. It's not too nice...

emilio reviewed (approved):
r=me. I think the test could be simpler, but it's probably not a huge deal I guess? (it'll surely be nicer though)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18003) made this pull request unmergeable. Please resolve the merge conflicts.

brainlessdeveloper commented:
r?

emilio commented:
@bors-servo r+

Thanks!

bors-servo commented:
:pushpin: Commit 14c5a1b has been approved by `emilio`

<!-- @bors-servo r=emilio 14c5a1b8d3dee7cfc8dfeb9314a0178c33197145 -->

bors-servo commented:
:hourglass: Testing commit 14c5a1b8d3dee7cfc8dfeb9314a0178c33197145 with merge 8a48578a2601769d61ab6f8429e7b1ecb19cbb84...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8259), [arm32](http://build.servo.org/builders/arm32/builds/8483), [arm64](http://build.servo.org/builders/arm64/builds/8516), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8601), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5365), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5399), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8599), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1075), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1078), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5299), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5357), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1077), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1081), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3841)
Approved by: emilio
Pushing 8a48578a2601769d61ab6f8429e7b1ecb19cbb84 to master...
